pyro/contrib/funsor/handlers/enum_messenger.py

```diff
@@ -0,0 +1,148 @@
+
+
+@_get_support_value.register(funsor.cnf.Contraction)
+def _get_support_value_contraction(funsor_dist, name, **kwargs):
+    delta_terms = [v for v in funsor_dist.terms
+                   if isinstance(v, funsor.delta.Delta) and name in v.fresh]
+    assert len(delta_terms) == 1
+    return _get_support_value(delta_terms[0], name, **kwargs)
+
+
+@_get_support_value.register(funsor.delta.Delta)
+def _get_support_value_delta(funsor_dist, name, **kwargs):
+    assert name in funsor_dist.fresh
+    return OrderedDict(funsor_dist.terms)[name][0]
+
+
+@_get_support_value.register(funsor.Tensor)
+def _get_support_value_tensor(funsor_dist, name, **kwargs):
+    assert name in funsor_dist.inputs
+    return funsor.Tensor(
+        funsor.ops.new_arange(funsor_dist.data, funsor_dist.inputs[name].size),
+        OrderedDict([(name, funsor_dist.inputs[name])]),
+        funsor_dist.inputs[name].size
+    )
+
+
+@_get_support_value.register(funsor.distribution.Distribution)
+def _get_support_value_distribution(funsor_dist, name, expand=False):
+    assert name == funsor_dist.value.name
+    return funsor_dist.enumerate_support(expand=expand)
+
+
+def _enum_strategy_full(dist, msg):
+    sample_dim_name = "{}__PARTICLES".format(msg['name'])
+    sample_inputs = OrderedDict({sample_dim_name: funsor.bint(msg["infer"]["num_samples"])})
+    sampled_dist = dist.sample(msg['name'], sample_inputs)
+    return sampled_dist
+
+
+def _enum_strategy_exact(dist, msg):
+    if isinstance(dist, funsor.Tensor):
+        dist = dist - dist.reduce(funsor.ops.logaddexp, msg['name'])
+    return dist
+
+
+def enumerate_site(dist, msg):
+    # TODO come up with a better dispatch system for enumeration strategies
+    if msg["infer"].get("num_samples", None) is None:
+        return _enum_strategy_exact(dist, msg)
+    elif msg["infer"]["num_samples"] > 1 and \
+            (msg["infer"].get("expand", False) or msg["infer"].get("tmc") == "full"):
+        return _enum_strategy_full(dist, msg)
+    raise ValueError("{} not valid enum strategy".format(msg))
+
+
+class EnumMessenger(NamedMessenger):
+    """
+    This version of EnumMessenger uses to_data to allocate a fresh enumeration dim
+    for each discrete sample site.
+    """
+    def _pyro_sample(self, msg):
+        if msg["done"] or msg["is_observed"] or msg["infer"].get("enumerate") != "parallel" \
+                or isinstance(msg["fn"], _Subsample):
+            return
+
+        if "funsor" not in msg:
+            msg["funsor"] = {}
+
+        unsampled_log_measure = to_funsor(msg["fn"], output=funsor.reals())(value=msg['name'])
+        msg["funsor"]["log_measure"] = enumerate_site(unsampled_log_measure, msg)
+        msg["funsor"]["value"] = _get_support_value(
+            msg["funsor"]["log_measure"], msg["name"], expand=msg["infer"].get("expand", False))
+        msg["value"] = to_data(msg["funsor"]["value"])
+        msg["done"] = True
+
+
+def queue(fn=None, queue=None,
```

Comment:
This function is copied verbatim from `pyro.poutine.handlers.queue`, but with `trace` and `replay` replaced by their `contrib.funsor`-specific counterparts added in this PR.